src/main/java/com/cleanroommc/groovyscript/compat/mods/artisanworktables/Tables.java

```diff
@@ -0,0 +1,458 @@
+        }
+
+        @RecipeBuilderMethodDescription
+        public ArtisanRecipeBuilder matrix(List<List<IIngredient>> matrix) {
+            if (!this.shaped) {
+                issues.add("Input matrix can only be used for shaped recipes");
+                return this;
+            }
+
+            ((ShapedGridBuilder) this.gridBuilder).matrix(matrix);
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription(field = "matrix")
+        public ArtisanRecipeBuilder shape(List<List<IIngredient>> matrix) {
+            return matrix(matrix);
+        }
+
+        @RecipeBuilderRegistrationMethod
+        @Override
+        public IArtisanRecipe register() {
+            if (!validate()) return null;
+
+            ArtisanRecipeData data = this.gridBuilder.register();
+            if (data == null) return null;
+
+            EnumTier tier = RecipeTierCalculator.calculateTier(
+                    this.registryName,
+                    data.width,
+                    data.height,
+                    this.tools.size(),
+                    this.input.size(),
+                    this.fluidInput.isEmpty() ? null : this.fluidInput.get(0)
+            );
+            // weird cornercase that the mod doesn't handle
+            if (!shaped && data.ingredients.size() > 9 && tier != null) {
+                if (data.ingredients.size() <= 25) tier = EnumTier.WORKSHOP;
+                else tier = null;
+            }
+            GroovyLog.Msg message =
+                    GroovyLog.msg("Error adding Artisan Worktables recipe").error()
+                    .add(tier == null, "Unable to calculate the table tier");
+            if (tier != null) {
+                int tierNeeded = tier.getId();
+                message.add(tierNeeded > maximumTier, "The recipe requires at least tier {}, but maximumTier was set to {}", tierNeeded, maximumTier);
+                minimumTier = Math.max(minimumTier, tierNeeded);
+            }
+            if (message.postIfNotEmpty()) return null;
+
+            // The mod expects three Nonnull ExtraOutputChancePairs... yuck
+            ExtraOutputChancePair[] extraOutputs = new ExtraOutputChancePair[3];
+            Arrays.fill(extraOutputs, new ExtraOutputChancePair(ArtisanItemStack.EMPTY, 0));
+            for (int i = 0; i < this.extraOutputs.size(); i++) extraOutputs[i] = this.extraOutputs.get(i);
+
+            IArtisanRecipe recipe = new ArtisanRecipe(
+                    this.makeArtisanName(),
+                    requirements,
+                    output,
+                    tools.toArray(new ToolEntry[0]),
+                    data.ingredients.stream().map(r -> ArtisanIngredient.from(r.toMcIngredient())).collect(Collectors.toList()),
+                    input.stream().map(r -> ArtisanIngredient.from(r.toMcIngredient())).collect(Collectors.toList()),
+                    consumeSecondaryIngredients,
+                    fluidInput.isEmpty() ? null : fluidInput.get(0),

```

Comment:
`fluidInput.getOrDefault(0, null)` works, right?

Reply:
no because it's not a list